### ISSUE TITLE
vmagent: check config option to verify scrape_configs / relabel_config files

### ISSUE THREAD

**Issue by iain-buclaw-sociomantic** [VictoriaMetrics/VictoriaMetrics#362]
**Is your feature request related to a problem? Please describe.**
A way to verify config passed to vmagent won't be rejected on start-up, intended as a step to run before starting or restarting the vmagent instance.

Currently, only way to find if new config will be accepted is to run vmagent and check for the message, e.g:
```
cannot parse Prometheus config from "/path/to/config": cannot unmarshal data: yaml: line 4: mapping values are not allowed in this context
``` 

You _can_ use `promtool check config` for the `scrape_configs` file, but there no way to verify the `relabel_config` files.

**Describe the solution you'd like**
No preference.

**Describe alternatives you've considered**
There are a number of ways this could be done.
- Prometheus has a dedicated tool for checking config `promtool check config /path/to/config` (I see there is a vmctl tool repository).
- HAproxy has a command-line option `haproxy -c -f /path/to/config`.

**Additional context**
vmagent is running on a bare metal server, service and config files managed by puppet.

**Comment by valyala** [comment]
@iain-buclaw-sociomantic , thanks for the feature request. I think it can be implemented inside [vmctl](https://github.com/VictoriaMetrics/vmctl) tool in the way similar to `promtool`.
@hagen1778 , @tenmozes , what do you think about this?

**Comment by hagen1778** [comment]
The feature request seems reasonable. However, I'm finding it a bit strange to install one tool to check the configuration for another tool. Let's imagine some work scenarios:
1. The rules are stored in repository and user wants to verify the changes when pushing a branch. This will require CI to install a vmctl tool in order to run validation. Basically, the same could be done by installing vmagent tool in CI env
2. VMagent runs in k8s and user wants to update the rules. He does some changes and deploys the manifest, k8s will spin-up a new pod which will immediately fail with config error. The health-checks will fail and old healthy pod will continue to work.
3. VMagent runs on a bare metal. User applies changes to rules, deploys the file and sends SIGHUP. VMagent won't accept the new config, log an error and continue to work.

The way how it was done in Prometheus doesn't mean that we should do the same. This is just my opinion and I'd be glad to hear other mentioned guys.

Btw, the error you linked `mapping values are not allowed in this context` could be caught with any yaml linter, which is the common practice to use when working with yaml-configs.

**Comment by iain-buclaw-sociomantic** [comment]
> Btw, the error you linked mapping values are not allowed in this context could be caught with any yaml linter, which is the common practice to use when working with yaml-configs.

It was a deliberate syntax error I put by way of example, of course there are better ways to get out errors from vmagent such as:
```
---
global:
  scrape_interval: 30t
  evaluation_interval: 30s
```
Which is valid yaml, but gets us.
```
cannot read "/etc/victoria/agent.yml": cannot parse Prometheus config from "/etc/victoria/agent.yml": cannot unmarshal data: yaml: unmarshal errors:
  line 3: cannot unmarshal !!str `30t` into time.Duration
```

> VMagent runs on a bare metal. User applies changes to rules, deploys the file and sends SIGHUP. VMagent won't accept the new config, log an error and continue to work.

Until the server reboots and vmagent doesn't start up at all. :-)

There are alternative ways to check file contents, and the config itself is not expected to change much on my end, plus a mixture of peer review/testing measures in place should ensure that an accidental outage shouldn't happen anyway, so this is pretty low priority for me.

The only use-case envisioned was to manage vmagent config files from puppet in the following manner:
```
file { '/etc/victoria/agent.yml':
  content      => epp('victoria/etc/agent.yml.epp', {
    config => $config.to_yaml,
  }),
  validate_cmd => '/usr/bin/vmagent -check.config -promscrape.config=%',
}
```
Where 'validate_cmd' ensures that config changes are not committed proper unless the command returns zero.

**Comment by hagen1778** [comment]
Got your point! There are plans to add config check via additional flag, like in your last example. So keeping this issue as enhancement. Thanks for detailed responses!

**Comment by valyala** [comment]
FYI, `vmagent` already supports `-promscrape.config.strictParse` option, which emits errors when it finds unsupported options in `-promscrape.config` file. `vmagent` continues working if the config has no errors or unsupported options.

I'm going to add `-promscrape.config.dryRun` option, which will work similar to `-promscrape.config.strictParse` on config parse errors, but will exit with `0` status code on config parse success.

**Comment by valyala** [comment]
The `-promscrape.config.dryRun` option has been added in the commit d1a9d8aa1c938bdefb7fe68e8d226370513ba383 . This commit will be included in the next release.

**Comment by valyala** [comment]
Added also option `-dryRun` for `vmagent` in the commit f9847352b4b5fe6e810559e63681c680454a5367 . This option checks the following configs for errors:
* `-promscrape.config`
* `-remoteWrite.relabelConfig`
* `-remoteWrite.urlRelabelConfig`

**Comment by valyala** [comment]
The `-dryRun` option has been added into `vmagent` starting from [v1.35.6](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing this feature request as done.

Feel free re-opening it if the feature doesn't work as expected.
